### PR TITLE
CSP: Handle distinction between no policy and an empty policy correctly; refine some relevant policy checks

### PR DESCRIPTION
**Pull request by morlovich**

### CODE REVIEW

File: net/instaweb/rewriter/css_filter_test.cc [hunk 1]
```diff
@@ -2492,6 +2493,27 @@ TEST_F(CssFilterTest, RenderCsp) {
              "Content Security Policy-->"));
 }
 
+TEST_F(CssFilterTest, InlineCspIrrelevant) {
+  EnableDebug();
+  options()->ClearSignatureForTesting();
+  options()->EnableFilter(RewriteOptions::kRewriteStyleAttributes);
+  server_context()->ComputeSignature(options());
+
+  // We don't worry about not-style, not-default CSP, shough
```
Comment (oschaaf) [inline net/instaweb/rewriter/css_filter_test.cc]: s/sough/though/

**Review by oschaaf — APPROVED** [review]
I think this looks good, barring one typo.
(Also restarted the timed out Travis job)

File: net/instaweb/rewriter/javascript_filter_test.cc [hunk 2]
```diff
@@ -1468,10 +1468,28 @@ TEST_P(JavascriptFilterTest, RenderCsp) {
...
 
-TEST_P(JavascriptFilterTest, InlineCsp) {
+TEST_P(JavascriptFilterTest, CspIrrelevant) {
   InitFilters();
   EnableDebug();
 
+  SetResponseWithDefaultHeaders(
+      "scripts/a.js", kContentTypeJavascript, kJsData, 100);
+
+  const char kCsp[] =
```
Comment (jmarantz) [inline net/instaweb/rewriter/javascript_filter_test.cc]: static

File: net/instaweb/rewriter/javascript_filter_test.cc [hunk 3]
```diff
@@ -1486,6 +1504,42 @@ TEST_P(JavascriptFilterTest, InlineCsp) {
              "<!--Avoiding modifying inline script with CSP present-->"));
 }
 
+TEST_P(JavascriptFilterTest, InlineCsp2) {
+  InitFilters();
+  EnableDebug();
+
+  const char kCsp[] =
```
Comment (jmarantz) [inline net/instaweb/rewriter/javascript_filter_test.cc]: static

File: net/instaweb/rewriter/javascript_filter_test.cc [hunk 4]
```diff
@@ -1486,6 +1504,42 @@ TEST_P(JavascriptFilterTest, InlineCsp) {
...
+      StrCat(kCsp, kScript,
+             "<!--Avoiding modifying inline script with CSP present-->"));
+}
+
+TEST_P(JavascriptFilterTest, InlineCsp3) {
+  InitFilters();
+  EnableDebug();
+
+  // This one doesn't restrict scripts.
+  const char kCsp[] =
```
Comment (jmarantz) [inline net/instaweb/rewriter/javascript_filter_test.cc]: static

**Review by jmarantz — APPROVED** [review]

File: net/instaweb/rewriter/css_filter_test.cc (hunk 1 above)
Comment (morlovich) [reply net/instaweb/rewriter/css_filter_test.cc]: Done. Actually, just dropped the word.

File: net/instaweb/rewriter/csp.cc [hunk 5]
```diff
@@ -495,6 +495,14 @@ std::unique_ptr<CspPolicy> CspPolicy::Parse(StringPiece input) {
         policy->policies_[static_cast<int>(dir_name)]
             = CspSourceList::Parse(value);
       }
+    } else {
+      // Empty policy
+      CspDirective dir_name = LookupCspDirective(token);
+      if (dir_name != CspDirective::kNumSourceListDirectives &&
+          policy->policies_[static_cast<int>(dir_name)] == nullptr) {
```
[the remark this answers is not in the document]
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc]: CSE'd the static_cast instead; I think references are a little harder to reason about then a short array use...

File: net/instaweb/rewriter/csp_test.cc [hunk 6]
```diff
@@ -449,6 +449,12 @@ TEST(CspParseSourceListTest, None) {
   EXPECT_TRUE(n2->expressions().empty());
 }
 
+TEST(CspParseSourceListTest, Empty) {
+  std::unique_ptr<CspSourceList> l(CspSourceList::Parse(""));
```
[the remark this answers is not in the document]
Comment (morlovich) [reply net/instaweb/rewriter/csp_test.cc]: Changed.

File: net/instaweb/rewriter/csp_test.cc [hunk 7]
```diff
@@ -725,9 +745,9 @@ TEST(CspPolicyTest, CanLoadUrl) {
 
   {
     std::unique_ptr<CspPolicy> p(CspPolicy::Parse("child-src *"));
-    EXPECT_FALSE(p->CanLoadUrl(CspDirective::kImgSrc,
-                               GoogleUrl("http://www.example.com/"),
-                               GoogleUrl("http://www.example.org/foo.png")));
+    EXPECT_TRUE(p->CanLoadUrl(CspDirective::kImgSrc,
+                              GoogleUrl("http://www.example.com/"),
```
[the remark this answers is not in the document]
Comment (morlovich) [reply net/instaweb/rewriter/csp_test.cc:749]: I find this actually easier to read, and I like the lack of indirection --- no need to keep state about bindings in my brain.

File: net/instaweb/rewriter/css_filter_test.cc [hunk 8]
```diff
@@ -2492,6 +2493,27 @@ TEST_F(CssFilterTest, RenderCsp) {
...
+  const char kCsp[] = "<meta http-equiv=\"Content-Security-Policy\" "
```
[the remark this answers is not in the document]
Comment (morlovich) [reply net/instaweb/rewriter/css_filter_test.cc]: Done

File: net/instaweb/rewriter/javascript_filter_test.cc (hunk 2 above)
Comment (morlovich) [reply net/instaweb/rewriter/javascript_filter_test.cc]: Done

File: net/instaweb/rewriter/javascript_filter_test.cc (hunk 3 above)
Comment (morlovich) [reply net/instaweb/rewriter/javascript_filter_test.cc]: Done

File: net/instaweb/rewriter/javascript_filter_test.cc (hunk 4 above)
Comment (morlovich) [reply net/instaweb/rewriter/javascript_filter_test.cc]: Done

**Review by jmarantz — APPROVED** [review]

**Comment by oschaaf** [comment]
Restarted the travis job, which timed out. checkin tests succeeded:

http://ci.onpagespeed.com/stream/mod_pagespeed/0c9a9beb6295f6b96fa52f47f77c115d8cedaf0e/ubuntu-1404-x64-checkin-test

Other tests passing too:
http://ci.onpagespeed.com/builds/mod_pagespeed/f2dd281e4f7a22b3704278dccf5b2c506852cadc/

**Pull request merged**